Feature/#6 datepicker 패키지 추가

jm3789 opened the pull request:
## 🔍 변경 사항
- [x] 날짜 타입, defaults 요소(색상 등) 정의
- [x] DatePickerState 설계
  - [x] 선택된 날짜 기준으로 연/월 결정
- [x] ui 구성
  - [x] 날짜, 연/월 ui 전환
  - [x] 헤더 아이콘 추가
  - [x] 공통 Theme 적용
  - [x] Preview 추가
  - [x] 바텀시트 컴포넌트 추가
- [x] 지난 날짜 비활성화
- [x] configuration change 대응
- [x] 각 날짜 셀의 상태를 리스트로 관리

### 사용 방법

```kotlin
val datePickerState = rememberDatePickerState(
    initialSelectedDate = LocalDate(2025, 12, 31),  // 기본값: null
    initialMode = DatePickerMode.YEAR_MONTH,  // 기본값: DatePickerMode.DATE
)

DatePicker(
    state = datePickerState,
)

Text(text = "날짜: ${datePickerState.selectedDate?.toString()}")
```
- `rememberDatePickerState`로 상태를 생성한 뒤 `DatePicker` 컴포저블에 전달해 사용합니다. 
- 초기 선택 날짜(`initialSelectedDate`)와 시작 모드(`initialMode`, 달력 모드 or 연/월 모드)를 지정할 수 있고, 선택된 날짜는 `state.selectedDate`으로 확인할 수 있습니다.
- 날짜 타입은 `kotlinx.datetime.LocalDate`를 사용합니다.

### 패키지 구조
<img width="342" height="317" alt="스크린샷 2026-01-02 오후 7 14 36" src="https://github.com/user-attachments/assets/192beb54-0fbc-4325-a9b5-c8ed3b7125a5" />

## 📸 스크린샷 (선택)
[Screen_recording_20251230_004516.webm](https://github.com/user-attachments/assets/e5f1efbe-9db4-48d1-a6c9-5cb6892e3fdb)

**+) 날짜선택 바텀시트 ui 컴포넌트 추가**
<img width="180" height="320" alt="Screenshot_20260102_073220" src="https://github.com/user-attachments/assets/ee1f1b80-0f7a-4b99-894e-5eb2cc22e853" />

**+) 변경될 필요 없는 날짜 셀은 리컴포지션되지 않도록 개선**
<img width="488" height="506" alt="스크린샷 2026-01-02 오후 6 13 30-side" src="https://github.com/user-attachments/assets/db8ca555-69ac-41d0-82a5-4e3d5ef92e24" />



작업 과정을 [문서](https://shorthaired-mailman-881.notion.site/2d7c50288afe80108b77dd281f81329f?pvs=74)에 정리해두었습니다.

## 🔗 관련 이슈
- #6 

- Close #6

## ✅ 체크리스트
- [x] 팀 컨벤션을 준수했나요?
- [x] 불필요한 주석이나 import는 삭제했나요?

jm3789 committed:
build: kotlinx datetime 의존성 추가

jm3789 committed:
feat: DatePickerState 추가, 연/월 및 날짜 클래스 추가

jm3789 committed:
feat: DatePicker ui 컴포넌트 추가 (달력 / 연월)

jm3789 committed:
fix: DatePickerState 에서 DatePickerDate 대신 LocalDate만 노출하도록 수정

jm3789 committed:
fix: 어노테이션 수정

jm3789 committed:
chore: 불필요한 확장함수 제거

jm3789 committed:
feat: 헤더 양옆 화살표로 연/월 한 칸씩 이동

jm3789 committed:
fix: 달력이 일요일부터 시작하도록 요일인덱스 수정

jm3789 committed:
fix: Stable 어노테이션 추가

jm3789 committed:
feat: 오늘 날짜 회색 표시

jm3789 committed:
fix: 선택된 날짜가 있는 경우 해당 날짜의 달력을 우선표시

jm3789 committed:
Merge branch 'feature/#3-theme' of https://github.com/boostcampwm2025/and02-boostcamp into feature/#6-component-datepicker

jm3789 committed:
feat: 모드 전환 시 Crossfade 애니메이션 추가

jm3789 committed:
feat: 오늘보다 이전 날짜면 선택 불가하도록 추가

jm3789 committed:
feat: DatePickerDefaults.colors() 정의해 컴포저블에서 사용, InvitationTheme 색상 사용

jm3789 committed:
feat: Invitation theme 요소 적용

jm3789 committed:
fix: 오늘 날짜 배경색상 수정

jm3789 committed:
chore: 날짜 클릭 효과 제거

jm3789 committed:
chore: 아이콘 추가

jm3789 committed:
feat: 헤더에 아이콘 적용 및 클릭효과 제거

jm3789 committed:
fix: 패딩 수정

jm3789 committed:
feat: DatePickerDefaults에 today() 추가

jm3789 committed:
feat: today 계산하는 대신 defaults의 today() 사용

jm3789 committed:
chore: 사용하지 않는 import문 제거

jm3789 committed:
fix: state에 month 정보 안 들어가도록 수정

jm3789 committed:
feat: state에 saveable 추가

jm3789 committed:
fix: Saver 정보에 mode 추가

jm3789 committed:
feat: state에 initialMode 파라미터 추가

jm3789 committed:
Merge branch 'dev' of https://github.com/boostcampwm2025/and02-boostcamp into feature/#6-component-datepicker

# Conflicts:
#	android/core/designsystem/build.gradle.kts

jm3789 committed:
fix: 잘못 push한 코드 수정

jm3789 committed:
feat: DatePicker preview 추가

jm3789 committed:
chore: Ktlint 검사

jm3789 committed:
chore: lint 검사 후 수정

jm3789 committed:
fix: 날짜 행 간격 수정

seojw0124 commented on android/core/designsystem/src/main/kotlin/com/andlife/designsystem/component/datepicker/ui/InvitationDatePickerDefaults.kt:
```diff
@@ -0,0 +1,59 @@
+import com.andlife.designsystem.theme.InvitationTheme
+import kotlinx.datetime.Clock
+import kotlinx.datetime.LocalDate
+import kotlinx.datetime.TimeZone
+import kotlinx.datetime.todayIn
+
+@Stable
+object DatePickerDefaults {
+    fun today(
+        clock: Clock = Clock.System,
+        timeZone: TimeZone = TimeZone.currentSystemDefault(),
+    ): LocalDate = clock.todayIn(timeZone)
```
Default 값까지 만들어주셨군요. 좋습니다!

seojw0124 commented on android/core/designsystem/src/main/kotlin/com/andlife/designsystem/component/datepicker/model/DatePickerYearMonth.kt:
```diff
@@ -0,0 +1,38 @@
+            year = year,
+            month = month - 1,
+        )
+    }
```
이건 지금도 괜찮은 것 같고, DatePickerYearMonth의 멤버 함수로 넣어도 될 것 같기도 합니다.

```kotlin
data class DatePickerYearMonth(
    val year: Int,
    val month: Int,
) {
    fun plusMonth(): DatePickerYearMonth = 
        if (month == 12) ...

    fun minusMonth() 이런 식...
}
```

jm3789 replied:
그렇네요 둘다 `yearMonth.plusMonth()`형태로 호출할 수 있겠네요.
기존처럼 확장함수로 넣으면 함수 이름으로 따로 import 해야해서 멤버함수로 두는 게 더 좋을 것 같습니다. 수정해두겠습니다!

seojw0124 commented on android/core/designsystem/src/main/kotlin/com/andlife/designsystem/component/datepicker/model/InvitationDatePickerDate.kt:
```diff
@@ -0,0 +1,9 @@
+import androidx.compose.runtime.Immutable
+import kotlinx.datetime.LocalDate
+
+@Immutable
+data class DatePickerDate(
+    val date: LocalDate,
+)
```
저도 잘 모르긴 하는데, 동경님이 최상단에 stability_config.conf 파일에 LocalDate를 추가해둬서 이거 안 써도 되지 않을까요..?
내일 이야기해봐요!

jm3789 replied:
맞습니다!
-> stability_config에 추가하기 전에 이미 작성한 코드라서 일단 가져가는 것으로 논의했습니다

seojw0124 reviewed (commented):
고생하셨습니다!

jm3789 committed:
fix: 이름 앞에 Invitation 추가

jm3789 committed:
fix: plusMonth(), minusMonth() 멤버함수로 변경

dongykung commented on android/core/designsystem/src/main/kotlin/com/andlife/designsystem/component/datepicker/ui/InvitationDatePicker.kt:
```diff
@@ -0,0 +1,400 @@
+            InvitationDatePickerMode.DATE -> {
+                Column(modifier) {
+                    InvitationDatePickerHeader(
+                        title = "${state.displayedMonth.year}년 ${state.displayedMonth.month}월",
```
하드코딩보단 strings.xml 활용해주시면 감사하겠습니다!

jm3789 replied:
넵 수정했습니다!

dongykung commented on android/core/designsystem/src/main/kotlin/com/andlife/designsystem/component/datepicker/ui/InvitationDatePicker.kt:
```diff
@@ -0,0 +1,400 @@
+            InvitationDatePickerMode.YEAR_MONTH -> {
+                Column(modifier) {
+                    InvitationDatePickerHeader(
+                        title = "${state.displayedMonth.year}년",
```
마찬가지임다!

dongykung commented on android/core/designsystem/src/main/kotlin/com/andlife/designsystem/component/datepicker/ui/InvitationDatePicker.kt:
```diff
@@ -0,0 +1,400 @@
+    val firstDayOfWeek = getFirstDayOfWeek(yearMonth)
+
+    Column(modifier = modifier.padding(horizontal = InvitationSpacing.large)) {
+        // 요일 헤더
+        Row(
+            modifier = Modifier.fillMaxWidth(),
+            horizontalArrangement = Arrangement.SpaceBetween,
+        ) {
+            listOf("일", "월", "화", "수", "목", "금", "토").forEach { dayOfWeek ->
+                Text(
+                    text = dayOfWeek,
```
이 부분도 하드코딩말고 strings나 
```kotlin
DayOfWeek.entries.map {
        it.getDisplayName(TextStyle.NARROW, locale)
    }
``` 
Kotlinx.datetime에 해당 옵션을 사용해보셔도 좋을 것 같습니다

jm3789 replied:
이런 것도 있었군요, 이후에 "년"과 "월"에도 locale을 적용해볼 수 있을 것 같습니다.
일단 수정했는데, TextStyle에서 java.time 을 import하게 되었는데 괜찮을까요?

dongykung commented on android/core/designsystem/src/main/kotlin/com/andlife/designsystem/component/datepicker/ui/InvitationDatePicker.kt:
```diff
@@ -0,0 +1,400 @@
+            "10월",
+            "11월",
+            "12월",
+        )
```
요 부분도 위 사항 반영해주심 감사하겠습니다

dongykung commented on android/core/designsystem/src/main/kotlin/com/andlife/designsystem/component/datepicker/ui/InvitationDatePicker.kt:
```diff
@@ -0,0 +1,400 @@
+    }
+
+// 윤년 여부 판단
+private fun isLeapYear(year: Int): Boolean = year % 4 == 0 && (year % 100 != 0 || year % 400 == 0)
```
개인적인 의견입니다만 UI는 state에서 수행된 데이터를 보여주기만하고 
StateImpl 객체에서 수행되게 하는 걸 선호하긴 하는데 어덯게 생각하시나요?

jm3789 replied:
넵 말씀해주신대로 UI 에서는 표현만 해주는 게 좋을 것 같습니다.
이 부분은 월 값에 따른 계산로직이라 일단 state 대신 YearMonth 모델의 확장함수로 옮겨두었습니다.

dongykung commented on android/core/designsystem/src/main/kotlin/com/andlife/designsystem/component/datepicker/ui/InvitationDatePicker.kt:
```diff
@@ -0,0 +1,400 @@
+                    )
+                val isSelected = selectedDate?.date == date.date
+                val isToday = date.date == InvitationDatePickerDefaults.today()
+                val isBeforeToday = date.date < InvitationDatePickerDefaults.today()
```
이 부분도 stateImpl에서 객체로 관리하는건 어덯게 생각하시나요?

jm3789 replied:
이부분도 객체로 관리하는 게 좋을 것 같습니다. 다만 stateImpl에서 어떤 식으로 관리해야 할지 감이 잡히지 않아 ai에게 물어보았는데, StateImpl이 isSelected, isToday 등의 셀 단위 UI 상태를 직접 알게 되면 표현의 책임이 분산된다고 해서, 일단 `DateCellUiModel`만 따로 정의해서 진행했습니다. [커밋](https://github.com/boostcampwm2025/and02-boostcamp/pull/13/commits/d8b291b193461144fe44d459801e6f49797b66d0)
관련해 다른 방법이 있다면 피드백 주시면 감사하겠습니다!

CDHyun commented on android/core/designsystem/src/main/kotlin/com/andlife/designsystem/component/datepicker/ui/InvitationDatePicker.kt:
```diff
@@ -0,0 +1,400 @@
+    ) {
+        Icon(
+            painter = painterResource(id = R.drawable.ic_chevron_left_24),
+            contentDescription = null,
```
`contentDescription`도 문자열 리소스로 관리하여 채워주시면 좋을 것 같습니다!

jm3789 replied:
넵 수정했습니다!

jm3789 committed:
feat: datepicker 바텀시트 컴포넌트 추가

ant5555 commented on android/core/designsystem/src/main/kotlin/com/andlife/designsystem/component/datepicker/state/InvitationDatePickerState.kt:
```diff
@@ -0,0 +1,42 @@
+}
+
+@Composable
+fun rememberInvitationDatePickerState(
+    initialSelectedDate: LocalDate? = null,
+    initialMode: InvitationDatePickerMode = InvitationDatePickerMode.DATE,
+): InvitationDatePickerState =
+    rememberSaveable(saver = InvitationDatePickerStateImpl.Saver()) {
+        InvitationDatePickerStateImpl(
+            initialSelectedDate = initialSelectedDate?.let { InvitationDatePickerDate(it) },
+            initialMode = initialMode,
+        )
```
Configuration change 될때 선택된 날짜가 초기화되니까 Saver가 필요하겠군요
노션에 정리해주신 내용보면서 이해하면서 볼 수 있었습니다! 감사합니다

ant5555 commented on android/core/designsystem/src/main/kotlin/com/andlife/designsystem/component/datepicker/state/InvitationDatePickerStateImpl.kt:
```diff
@@ -0,0 +1,92 @@
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.saveable.Saver
+import androidx.compose.runtime.saveable.listSaver
+import com.andlife.designsystem.component.datepicker.model.InvitationDatePickerDate
+import com.andlife.designsystem.component.datepicker.model.InvitationDatePickerYearMonth
+import com.andlife.designsystem.component.datepicker.model.toYearMonth
+import com.andlife.designsystem.component.datepicker.ui.InvitationDatePickerDefaults
+import com.andlife.designsystem.component.datepicker.ui.InvitationDatePickerMode
+import kotlinx.datetime.LocalDate
+
+@Stable
+internal class InvitationDatePickerStateImpl(
+    initialSelectedDate: InvitationDatePickerDate?,
+    initialMode: InvitationDatePickerMode,
+) : InvitationDatePickerState {
+    private val _selectedDate = mutableStateOf(initialSelectedDate)
+    private val _displayedMonth =
+        mutableStateOf(
+            initialSelectedDate?.toYearMonth()
+                ?: InvitationDatePickerDate(InvitationDatePickerDefaults.today()).toYearMonth(),
+        )
+    private val _mode = mutableStateOf(initialMode)
+
+    override val selectedDate: LocalDate?
+        get() = _selectedDate.value?.date // InvitationDatePickerDate에서 LocalDate로 변환
+
+    override val displayedMonth: InvitationDatePickerYearMonth
+        get() = _displayedMonth.value
+
+    override val mode: InvitationDatePickerMode
+        get() = _mode.value
+
+    override fun selectDate(date: LocalDate) {
+        _selectedDate.value = InvitationDatePickerDate(date) // LocalDate에서 InvitationDatePickerDate로 변환
+    }
```
저도 지민님처럼
state 구현 내부에서는 커스텀 타입을 쓰고, 외부에는 LocalDate(표준타입)만 가지고도 사용할 수 있도록 해야하니까, state를 interface로 분리한 내용으로 이해했습니다.

ant5555 reviewed (approved):
고생하셨습니다! 문서로 잘 정리해주셔서 코드 이해하면서 읽기가 수월했습니다.

jm3789 committed:
chore: 바텀시트 하단 버튼 string resource 추가

jm3789 committed:
chore: string resource 추가, defaults에서 resource id 관리

jm3789 committed:
feat: InvitationDatePicker에 locale 파라미터 추가, locale에 맞게 달력에 요일 출력

- 기본값은 기기에 설정된 언어입니다.

jm3789 committed:
chore: contentDescription 추가 및 resource로 관리

jm3789 committed:
chore: 월 첫날 요일인덱스, 일수 반환 로직을 YearMonth의 확장함수로 이동

jm3789 committed:
chore: 달력의 날짜 cell을 나타내는 ui model 분리해서 구현

jm3789 committed:
Merge branch 'dev' into feature/#6-component-datepicker

dongykung commented on android/core/designsystem/src/main/kotlin/com/andlife/designsystem/component/datepicker/ui/InvitationDatePicker.kt:
```diff
@@ -225,29 +225,41 @@ private fun InvitationDatePickerCalendar(
-                val isSelected = selectedDate?.date == date.date
-                val isToday = date.date == InvitationDatePickerDefaults.today()
-                val isBeforeToday = date.date < InvitationDatePickerDefaults.today()
+                }
```
해당 UIModel을 UI에서 생성하고 있는데 이러한 로직 또한 State에서 할 수 있었으면 더 좋을 것 같습니다!

jm3789 committed:
fix: state에서 date cell 리스트를 계산하여 사용

jm3789 committed:
chore: 네이밍 변경

- InvitationDatePickerCellUiModel -> InvitationDatePickerCell

jm3789 committed:
fix: date cell의 key로 문자열을 사용

jm3789 committed:
chore: lint format

jm3789 committed:
fix: 달력 그리드에서 state.dateCells 바로 사용

- 기존 코드에서 월을 이동할 때 앱 강제종료가 발생함: 월을 변경할 때 dateCells 의 derivedStateOf가 재계산되면서, key 생성에서 오류가 발생하는 것으로 추정됨

jm3789 committed:
Merge branch 'dev' into feature/#6-component-datepicker
